src/webgpu/api/operation/queue/writeBuffer.spec.ts

```diff
@@ -1,11 +1,53 @@
-export const description = `
-TODO:
-- source.origin is unaligned
-- data given as ArrayBuffer
-- data given as each TypedArray variant
-`;
+export const description = 'Operation tests for GPUQueue.writeBuffer()';
 
+import { params, pbool, poptions } from '../../../../common/framework/params_builder.js';
 import { makeTestGroup } from '../../../../common/framework/test_group.js';
+import { range } from '../../../../common/framework/util/util.js';
 import { GPUTest } from '../../../gpu_test.js';
 
+const kTypedArrays = [
+  'Uint8Array',
+  'Uint16Array',
+  'Uint32Array',
+  'Int8Array',
+  'Int16Array',
+  'Int32Array',
+  'Float32Array',
+  'Float64Array',
+] as const;
+
 export const g = makeTestGroup(GPUTest);
+
+const kTestData = range<number>(16, i => i);
+
+g.test('array_types')
```

Comment:
Can we also have tests that the offset and size arguments are handled correctly? (for typed arrays they are in elements while they are in bytes for ArrayBuffer).

What about a 0-sized writeBuffer?

What about many writeBuffers on the same location? (to check barriers are done correctly), overlapping location but only on one word at the start or the end?

Any other tests you can think of?

Reply:
I added the tests you suggested: 
- using offset and size on TypedArrays and ArrayBuffers
- 0-sized writes to non-zero and zero sized buffers
- multiple overlapping writes with decreasing sizes

I added another tests with overlap in the middle. Although, I'm not sure that it adds much.

I also refactored the tests so that they calculate the expected result given a list of writes.